### PR TITLE
SOLR-17367 - org.apache.solr.cli.PostTool is ignoring -params option

### PR DESCRIPTION
**Pull request by bogdanstoik**
https://issues.apache.org/jira/browse/SOLR-17367  

# Description

Adding support for -params option in org.apache.solr.cli.PostTool

# Solution

I modified the postData method so that it checks and appends -params value to the solr request

# Tests

I implemented testRunCsvWithCustomSeparatorParam in org.apache.solr.cli.PostToolTest. The test creates a tsv and submits it to the test core. It then checks to see if the index contains exactly 1 document.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Comment by epugh** [comment]
thanks for the addition of the tests..  I will inline some comments....   one thing, did you run `gradlew tidy`?  that does some code formatting.

**Review by epugh — CHANGES_REQUESTED** [review]
Looks like some great progress.  A couple of questions!   When I add it to `CHANGES.txt`, how would you like to be credited?

File: solr/core/src/java/org/apache/solr/cli/PostTool.java [hunk 1]
```diff
@@ -311,6 +312,8 @@ public void runImpl(CommandLine cli) throws Exception {
 
     args = cli.getArgs();
 
+    params = cli.hasOption("params") ? cli.getOptionValue("params") : "";
```
[the remark this answers is not in the document]
Comment (bogdanstoik) [reply solr/core/src/java/org/apache/solr/cli/PostTool.java:315]: That looks way better

File: solr/core/src/test/org/apache/solr/cli/PostToolTest.java [hunk 2]
```diff
@@ -147,6 +147,43 @@ public void testRunWithCollectionParam() throws Exception {
...
+    withBasicAuth(CollectionAdminRequest.createCollection(collection, "conf1", 1, 1, 0, 0))
+        .processAndWait(cluster.getSolrClient(), 10);
+
+    File tsvDoc = File.createTempFile("temp", ".tsv");
+
+    FileWriter fw = new FileWriter(tsvDoc, StandardCharsets.UTF_8);
+    fw.write("1\tmytitle\n");
+    fw.close();
+
+    String[] args = {
```
[the remark this answers is not in the document]
Comment (bogdanstoik) [reply solr/core/src/test/org/apache/solr/cli/PostToolTest.java:166]: I haven't run gradle tidy after writing the testcase 🤦‍♂️

File: solr/core/src/test/org/apache/solr/cli/PostToolTest.java [hunk 3]
```diff
@@ -147,6 +147,43 @@ public void testRunWithCollectionParam() throws Exception {
     assertEquals("*:* found unexpected number of documents", expectedDocCount, numFound);
   }
 
+  @Test
+  public void testRunCsvWithCustomSeparatorParam() throws Exception {
+    final String collection = "testRunCsvWithCustomSeparatorParam";
+
+    // Provide the port for the PostTool to look up.
```
[the remark this answers is not in the document]
Comment (bogdanstoik) [reply solr/core/src/test/org/apache/solr/cli/PostToolTest.java:154]: To be honest, i just copied the testRunWithCollectionParam function and modified it. Have little to no experience in java, so i thougt it was safest.

File: solr/core/src/test/org/apache/solr/cli/PostToolTest.java [hunk 4]
```diff
@@ -147,6 +147,43 @@ public void testRunWithCollectionParam() throws Exception {
...
+
+    String[] args = {
+      "post", "-c", collection, "-credentials", SecurityJson.USER_PASS, "-params", "\"separator=%09&header=false&fieldnames=id,title_s\"", "-type", "text/csv", tsvDoc.getAbsolutePath()
+    };
+    assertEquals(0, runTool(args));
+
+    int numFound = 0;
+    int expectedDocCount = 1;
+
+    for (int idx = 0; idx < 100; ++idx) {
```
[the remark this answers is not in the document]
Comment (bogdanstoik) [reply solr/core/src/test/org/apache/solr/cli/PostToolTest.java:174]: same as above comment, i just copy-pasted and modified the testRunWithCollectionParam function

**Comment by bogdanstoik** [comment]
> thanks for the addition of the tests.. I will inline some comments.... one thing, did you run `gradlew tidy`? that does some code formatting.

I did not 🤦‍♂️.
I had a blast writing java for the first time in 10 years. I promise I'll try to contribute more to the project, as it's a key part in my job.

**Comment by bogdanstoik** [comment]
> Looks like some great progress. A couple of questions! When I add it to `CHANGES.txt`, how would you like to be credited?

Bostoi

Thank you

File: solr/core/src/test/org/apache/solr/cli/PostToolTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bogdanstoik) [reply solr/core/src/test/org/apache/solr/cli/PostToolTest.java:174]: looking through the code, i think this loop just tries to make sure the changes are commited to index. it "wants" to find the correct results, after all 😄

File: solr/core/src/test/org/apache/solr/cli/PostToolTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bogdanstoik) [reply solr/core/src/test/org/apache/solr/cli/PostToolTest.java:166]: i've ran it just now and it does wrap

**Comment by bogdanstoik** [comment]
> > thanks for the addition of the tests.. I will inline some comments.... one thing, did you run `gradlew tidy`? that does some code formatting.
> 
> I did not 🤦‍♂️. I had a blast writing java for the first time in 10 years. I promise I'll try to contribute more to the project, as it's a key part in my job.

The only thing i'm not sure about is passing the params to the /extract component of the webMode.

**Comment by epugh** [comment]
One more thought which is that sometimes I find it easier to write a bats test instead of Java...   post.bats is used to test out the cli against a live solr...

**Comment by epugh** [comment]
by the way @bogdanstoik how did you find this bug?   Did you have a scripte using the `-params`?

**Comment by bogdanstoik** [comment]
> by the way @bogdanstoik how did you find this bug? Did you have a scripte using the `-params`?

Yes. I was trying to load a tab-separated csv and it kept complaining about not finding the right columns. I ended up using curl to post my file directly to the update handler, then came back to check if's something in the tool.

**Review by epugh — APPROVED** [review]

**Pull request merged**